File: frontend/src/lib/api/org-structure.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { fetchJson } from "@/lib/api/http";
import type { components } from "@/types/api.generated";

export type OfficeResponse = components["schemas"]["OfficeResponse"];
export type OfficeCreate = components["schemas"]["OfficeCreate"];
export type OfficeUpdate = components["schemas"]["OfficeUpdate"];

export type PostResponse = components["schemas"]["PostResponse"];
export type PostCreate = components["schemas"]["PostCreate"];
export type PostUpdate = components["schemas"]["PostUpdate"];

export type OfficeJurisdiction = OfficeCreate["jurisdiction"];

export const orgStructureQueryKeys = {
	all: () => ["org-structure"] as const,
	offices: () => ["org-structure", "offices"] as const,
	posts: () => ["org-structure", "posts"] as const,
};

export function listOffices() {
	return fetchJson<OfficeResponse[]>("/api/offices");
}

export function createOffice(body: OfficeCreate) {
	return fetchJson<OfficeResponse>("/api/offices", {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});
}

export function updateOffice(officeId: string, body: OfficeUpdate) {
	return fetchJson<OfficeResponse>(`/api/offices/${officeId}`, {
		method: "PATCH",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});
}

export function listPosts() {
	return fetchJson<PostResponse[]>("/api/posts");
}

export function createPost(body: PostCreate) {
	return fetchJson<PostResponse>("/api/posts", {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});
}

export function updatePost(postId: string, body: PostUpdate) {
	return fetchJson<PostResponse>(`/api/posts/${postId}`, {
		method: "PATCH",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});
}

export function useOfficesList() {
	return useQuery({
		queryKey: orgStructureQueryKeys.offices(),
		queryFn: listOffices,
	});
}

export function useCreateOffice() {
	const queryClient = useQueryClient();
	return useMutation({
		mutationFn: createOffice,
		onSuccess: () => {
			void queryClient.invalidateQueries({ queryKey: orgStructureQueryKeys.offices() });
		},
	});
}

export function useUpdateOffice() {
	const queryClient = useQueryClient();
	return useMutation({
		mutationFn: ({ officeId, body }: { officeId: string; body: OfficeUpdate }) =>
			updateOffice(officeId, body),
		onSuccess: () => {
			void queryClient.invalidateQueries({ queryKey: orgStructureQueryKeys.offices() });
		},
	});
}

export function usePostsList() {
	return useQuery({
		queryKey: orgStructureQueryKeys.posts(),
		queryFn: listPosts,
	});
}

export function useCreatePost() {
	const queryClient = useQueryClient();
	return useMutation({
		mutationFn: createPost,
		onSuccess: () => {
			void queryClient.invalidateQueries({ queryKey: orgStructureQueryKeys.posts() });
		},
	});
}

export function useUpdatePost() {
	const queryClient = useQueryClient();
	return useMutation({
		mutationFn: ({ postId, body }: { postId: string; body: PostUpdate }) =>
			updatePost(postId, body),
		onSuccess: () => {
			void queryClient.invalidateQueries({ queryKey: orgStructureQueryKeys.posts() });
		},
	});
}

File: backend/app/api/routes/org_structure.py
"""Organization-structure master data routes."""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, status

from app.api.deps import Session, TenantCtx, require_capability
from app.auth.principal import AuthPrincipal
from app.schemas.org_structure import (
    OfficeCreate,
    OfficeResponse,
    OfficeUpdate,
    PostCreate,
    PostResponse,
    PostUpdate,
)
from app.services.org_structure import (
    create_office,
    create_post,
    list_offices,
    list_posts,
    post_to_response,
    update_office,
    update_post,
)

router = APIRouter(tags=["org-structure"])


@router.post("/offices", response_model=OfficeResponse, status_code=status.HTTP_201_CREATED)
async def create_office_route(
    body: OfficeCreate,
    ctx: TenantCtx,
    db: Session,
    _principal: Annotated[AuthPrincipal, Depends(require_capability("manage_master_data"))],
) -> OfficeResponse:
    return await create_office(
        db,
        UUID(ctx.organization_id),
        name=body.name,
        jurisdiction=body.jurisdiction,
    )


@router.get("/offices", response_model=list[OfficeResponse])
async def list_offices_route(
    ctx: TenantCtx,
    db: Session,
    _principal: Annotated[AuthPrincipal, Depends(require_capability("view_master_data"))],
) -> list[OfficeResponse]:
    return await list_offices(db, UUID(ctx.organization_id))


@router.patch("/offices/{office_id}", response_model=OfficeResponse)
async def update_office_route(
    office_id: UUID,
    body: OfficeUpdate,
    ctx: TenantCtx,
    db: Session,
    _principal: Annotated[AuthPrincipal, Depends(require_capability("manage_master_data"))],
) -> OfficeResponse:
    return await update_office(
        db,
        UUID(ctx.organization_id),
        office_id,
        name=body.name,
        jurisdiction=body.jurisdiction,
    )


@router.post("/posts", response_model=PostResponse, status_code=status.HTTP_201_CREATED)
async def create_post_route(
    body: PostCreate,
    ctx: TenantCtx,
    db: Session,
    _principal: Annotated[AuthPrincipal, Depends(require_capability("manage_master_data"))],
) -> PostResponse:
    post = await create_post(
        db,
        UUID(ctx.organization_id),
        designation=body.designation,
        class_name=body.class_name,
    )
    return post_to_response(post)


@router.get("/posts", response_model=list[PostResponse])
async def list_posts_route(
    ctx: TenantCtx,
    db: Session,
    _principal: Annotated[AuthPrincipal, Depends(require_capability("view_master_data"))],
) -> list[PostResponse]:
    posts = await list_posts(db, UUID(ctx.organization_id))
    return [post_to_response(post) for post in posts]


@router.patch("/posts/{post_id}", response_model=PostResponse)
async def update_post_route(
    post_id: UUID,
    body: PostUpdate,
    ctx: TenantCtx,
    db: Session,
    _principal: Annotated[AuthPrincipal, Depends(require_capability("manage_master_data"))],
) -> PostResponse:
    post = await update_post(
        db,
        UUID(ctx.organization_id),
        post_id,
        designation=body.designation,
        class_name=body.class_name,
    )
    return post_to_response(post)
